shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -457,56 +489,66 @@ public Surface getSurface() {
     @Override
     @TargetApi(29)
     public Image acquireLatestImage() {
-      Image r;
+      PerImage r;
+      PerImage toClose;
       synchronized (this) {
-        r = this.latestImage;
-        this.latestImage = null;
+        r = this.lastProducedImage;
+        this.lastProducedImage = null;
+        toClose = this.lastConsumedImage;
+        this.lastConsumedImage = r;
       }
-      maybeWaitOnFence(r);
-      return r;
+      if (toClose != null) {
+        toClose.close();
+      }
+      if (r == null) {
+        return null;
+      }
+      maybeWaitOnFence(r.image);
+      return r.image;
     }
 
-    private void maybeMarkReaderForClose() {
+    private void maybeCloseReader(ImageReader reader) {
       synchronized (this) {
-        if (this.toBeClosedReader != null) {
-          // We only ever have two readers:
-          // 1) The reader to be closed after the next image is produced.
-          // 2) The reader being used to produce images.
+        if (this.lastConsumedImage != null && this.lastConsumedImage.reader == reader) {
+          // There is still a consumed image in flight for this reader. Don't close.
           return;
         }
-        this.toBeClosedReader = this.activeReader;
-        this.activeReader = null;
-      }
-    }
-
-    private void maybeCloseReader() {
-      if (this.toBeClosedReader == null) {
-        return;
+        if (!readersToClose.contains(reader)) {
+          return;
+        }
+        readersToClose.remove(reader);
       }
-      this.toBeClosedReader.close();
-      this.toBeClosedReader = null;
+      // Close the reader.
+      reader.close();
     }
 
     private void maybeCreateReader() {
-      if (this.activeReader != null) {
-        return;
+      synchronized (this) {
+        if (this.activeReader != null) {
+          return;
+        }
+        this.activeReader = createImageReader();
       }
-      this.activeReader = createImageReader();
     }
 
     /** Invoked for each method that is available. */
-    private void onImage(Image image) {
+    private void onImage(PerImage image) {
       if (released) {
         return;
       }
-      Image toClose;
+      PerImage toClose;
       synchronized (this) {
-        toClose = this.latestImage;
-        this.latestImage = image;
+        if (this.readersToClose.contains(image.reader)) {
+          Log.e(TAG, "Skipped frame because resize is in flight.");
+          image.close();
+          return;
+        }
+        toClose = this.lastProducedImage;
+        this.lastProducedImage = image;
       }
       // Close the previously pushed buffer.
       if (toClose != null) {
-        Log.e(TAG, "RawSurfaceTexture frame was not acquired in a timely manner.");
+        Log.e(TAG, "Dropped frame.");
```

Comment:
Style guide says log messages should be actionable by the user. What action should be taken here? Is this only useful to engine developers? If so consider lowering the verbosity.

Reply:
Done.